source/agora/utils/HumanReadable.d

```diff
@@ -0,0 +1,284 @@
+        enum EndFrom    = Hash.StringBufferSize - 4;
+        size_t count;
+        scope void delegate(scope const(char)[]) @safe wrapper = (scope data) @safe {
+                if (count < StartUntil)
+                {
+                    sink(data);
+                    if (count + data.length >= StartUntil)
+                        sink("...");
+                }
+                if (count >= EndFrom)
+                    sink(data);
+                count += data.length;
+            };
+        this.value.toString(wrapper);
+    }
+}
+
+@safe unittest
+{
+    static immutable Hash SomeHash =
+        "0x000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f"
+        ~ "000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f";
+    assert(format("%s", HashFmt(SomeHash)) == "0x0000...e26f");
+}
+
+/// Formatting struct for `PublicKey`
+private struct PubKeyFmt
+{
+    private const(PublicKey) value;
+
+    public this (ref const PublicKey r) @safe
+    {
+        this.value = r;
+    }
+
+    public void toString (scope void delegate(scope const(char)[]) @safe sink) @safe
+    {
+        // Public keys are 56 characters, only take the first and last 4
+        // Only format `0xABCD..EFGH`
+        enum StartUntil = 4;
+        enum EndFrom    = 56 - 4;
+        size_t count;
+        scope void delegate(scope const(char)[]) @safe wrapper = (scope data) @safe {
+                if (count < StartUntil)
+                {
+                    sink(data);
+                    if (count + data.length >= StartUntil)
+                        sink("...");
+                }
+                if (count >= EndFrom)
+                    sink(data);
+                count += data.length;
+            };
+        this.value.toString(wrapper);
+    }
+}
+
+@safe unittest
+{
+    PublicKey pubkey;
+    assert(format("%s", PubKeyFmt(pubkey)) == "GAAA...AWHF");
+
+}
+
+/// Formatting struct for `Input`
+private struct InputFmt
+{
+    private const(Input) value;
+
+    public this (ref const Input r) @safe
+    {
+        this.value = r;
+    }
+
+    public void toString (scope void delegate(scope const(char)[]) @safe sink) @safe
+    {
+        formattedWrite(sink, "%s[%d]:%s",
+            HashFmt(this.value.previous), this.value.index,
+            HashFmt(this.value.signature));
+    }
+}
+
+@safe unittest
+{
+    Input input;
+    assert(format("%s", InputFmt(input)) == "0x0000...0000[0]:0x0000...0000");
+}
+
+/// Formatting struct for `Output`
+private struct OutputFmt
+{
+    private const(Output) value;
+
+    public this (ref const Output r) @safe
+    {
+        this.value = r;
+    }
+
+    public void toString (scope void delegate(scope const(char)[]) @safe sink) @safe
+    {
+        formattedWrite(sink, "%s(%s)",
+            AmountFmt(this.value.value), PubKeyFmt(this.value.address));
+    }
+}
+
+@safe unittest
+{
+    Output output;
+    assert(format("%s", OutputFmt(output)) == "0.0000,000(GAAA...AWHF)");
```

Comment:
Hmm.. I would prefer Address(Value), ordering by importance from left to right. What do you think?

Reply:
SGTM